### ISSUE TITLE
Add basic rewrite engine framework

### ISSUE THREAD

**Issue by aalexandrov** [MaterializeInc/materialize#9143]

**Comment by aalexandrov** [comment]
I added [#10252]() as a fast follow-up PR to address [this comment](https://github.com/MaterializeInc/materialize/pull/10109#discussion_r790667207) from #10109.

**Comment by aalexandrov** [comment]
Closed after #10252 is merged.
